lib/solvers/LongDistancePairSolver/LongDistancePairSolver.ts

```diff
@@ -0,0 +1,170 @@
+    for (const chip of inputProblem.chips) {
+      this.chipMap[chip.chipId] = chip
+      for (const pin of chip.pins) {
+        pinMap.set(pin.pinId, { ...pin, chipId: chip.chipId })
+      }
+    }
+
+    for (const netId of Object.keys(netConnMap.netMap)) {
+      const allPinIdsInNet = netConnMap.getIdsConnectedToNet(netId)
+      const unconnectedPinIds = allPinIdsInNet.filter(
+        (pinId) => !connectedPinIds.has(pinId),
+      )
+
+      if (unconnectedPinIds.length < 2) continue
+
+      const unconnectedPins = unconnectedPinIds
+        .map((pinId) => pinMap.get(pinId)!)
+        .filter(Boolean)
+
+      this.queuedCandidatePairs.push(...getAllPossiblePairs(unconnectedPins))
+    }
+  }
+
+  override getConstructorParams() {
+    return this.params
+  }
+
+  override _step() {
+    // 1. Check if a sub-solver has finished
+    if (this.subSolver?.solved) {
+      const newTracePath = this.subSolver.solvedTracePath
+      if (newTracePath && this.currentCandidatePair) {
+        // Post-Validation: Check for collisions against other traces
+        const isTraceClear = !doesTraceOverlapWithExistingTraces(
+          newTracePath,
+          this.params.solvedTraces,
+        )
+
+        if (isTraceClear) {
+          const [p1, p2] = this.currentCandidatePair
+          const globalConnNetId = this.netConnMap.getNetConnectedToId(p1.pinId)!
+          const mspPairId = `${p1.pinId}-${p2.pinId}`
+
+          this.solvedLongDistanceTraces.push({
+            // MspConnectionPair properties
+            mspPairId,
+            dcConnNetId: globalConnNetId, // Using global as a stand-in
+            globalConnNetId,
+            pins: [p1, p2],
+
+            // SolvedTracePath specific properties
+            tracePath: newTracePath,
+            mspConnectionPairIds: [mspPairId],
+            pinIds: [p1.pinId, p2.pinId],
+          })
+        }
+      }
+      this.subSolver = null
+      this.currentCandidatePair = null
+    } else if (this.subSolver?.failed) {
+      this.subSolver = null
+      this.currentCandidatePair = null
+    }
+
+    // 2. If a sub-solver is running, let it continue
+    if (this.subSolver) {
+      this.subSolver.step()
+      return
+    }
+
+    // 3. If no sub-solver is active, start a new one
+    if (this.queuedCandidatePairs.length === 0) {
+      this.solved = true
+      return
+    }
+
+    this.currentCandidatePair = this.queuedCandidatePairs.shift()!
+    if (!this.currentCandidatePair) {
+      this.solved = true
+      return
+    }
+
+    this.subSolver = new SchematicTraceSingleLineSolver2({
+      inputProblem: this.params.inputProblem,
+      pins: this.currentCandidatePair,
+      chipMap: this.chipMap,
+    })
+  }
+
+  override visualize() {
+    const graphics = visualizeInputProblem(this.inputProblem)
+
+    for (const trace of this.solvedLongDistanceTraces) {
+      graphics.lines!.push({
+        points: trace.tracePath,
+        strokeColor: "purple",
+      })
+    }
```

Comment:
i think you should draw candidates as well to improve the visualization, maybe with a stroke dash (IMO, optional)